docs/data-deduplication.md

```diff
@@ -0,0 +1,121 @@
+data stream into equal, fixed-size chunks, referred to as Fixed-Size Chunking (FSC). Content-Defined Chunking (CDC)
+based approaches are proposed to address the boundary-shift problem faced by the FSC approach. Specifically, CDC
+declares chunk boundaries based on the byte contents of the data stream instead of on the byte offset, as in FSC,
+and thus helps detect more redundancy for deduplication. According to some recent studies, CDCbased deduplication
+approaches are able to detect about 10-20 percent more redundancy than the FSC approach.
+```
+
+So, there are file-level deduplication, Fixed-Size Chunking deduplication and Content-Defined Chunking deduplication.
+
+### File-Level Deduplication
+File-level deduplication watches for multiple copies of the same file, stores the first copy, and then just links the other references to the first file. Only one copy gets stored on the disk/tape archive.
+Ultimately, the space you save on disk relates to how many copies of the file there were in the file system.
+File-level deduplication has both pros and cons:
+- Pro: smaller index data. Indexes for file-level deduplication are significantly smaller, which takes less computational time/memory resource when duplicates are being determined.
+- Pro: less computation resource. File-level processes require less processing power due to the smaller index and reduced number of comparisons.
+- Con: lower deduplication rate. A change within the file causes the whole file to be saved again.
+- Con: not friend to on demand downloading. The whole bigger file may need to be downloaded to access any byte of data.
+- Con: not friend to Container Registry and CDN for data cache and garbage-collection.
+
+### Fixed-Size Chunking Deduplication
+The Fixed-Size Chunking deduplication divides file content into fixed-size chunks/slices, and deduplicates among those data chunks.
+It's a balanced solution with:
+- Mediate deduplication rate, higher than file-level deduplication, lower than Content-Define Chunking(CDC).
+- Mediate index data size, bigger than file-level deduplication, smaller than CDC.
+- Mediate CPU/memory resource consumption when accessing the deduplicated data.
+- Pro: friend to on demand loading and random access.
+- Pro: friend to implement in-kernel filesystem.
+- Pro: friend to Container Registry for garbage-collection.
+- Con: not friend to CDN for data cache.
+
+### Content-Defined Chunking Deduplication
+Content-defined chunking (CDC) is a method to split files into variable length chunks, where the cut points are defined by some internal features of the files.
+Unlike fixed-length chunks, variable-length chunks are more resistant to byte shifting.
+Thus, it increases the probability of finding duplicate chunks within a file and between files.
+However, CDC algorithms require additional computation to find the cut points which might be computationally expensive for some applications.
```

Comment:
It also increases hashing/index store overhead since the slices are too thick.

Reply:
Fixed